### ISSUE TITLE
[FEATURE] Metadata-only library

### ISSUE THREAD

**Issue by aaronleopold** [stumpapp/stump#288]
**Is your feature request related to a problem? Please describe.**

A library that ignores the underlying folder structure and instead uses metadata to construct itself and its series

**Additional context**

Originally discussed in https://github.com/stumpapp/stump/issues/254

**Comment by vuhe** [comment]
Is this feature only using metadata information to build a database? I have a habit of organizing manga series into a folder and creating metadata files (such as `info.toml` files) in this folder. Can this feature be directly read and converted into metadata?

By the way, is there a plan to support metadata plugins? I am currently using LANraragi, which is very convenient for obtaining e-hentai metadata, but has poor support for comic series.

**Comment by aaronleopold** [comment]
> Is this feature only using metadata information to build a database? 

Basically, yep! That is the idea for the feature.

> I have a habit of organizing manga series into a folder and creating metadata files (such as `info.toml` files) in this folder. Can this feature be directly read and converted into metadata?

Stump doesn't currently support toml metadata files, but you're welcome to create a feature request for it. If the schema deviates from what is typically found in EPUBs or ComicInfo.xml files, please outline it if you're able to or provide a link to an example.

> By the way, is there a plan to support metadata plugins? I am currently using LANraragi, which is very convenient for obtaining e-hentai metadata, but has poor support for comic series.

This is trickier to answer. I'm not familiar with LANraragi and am not by a computer atm, but in general a plugin system would not be trivial to implement. That's assuming it needs to function as a plugin, and not some other service that just runs in parallel operating on metadata within your library.

Feel free to create a separate issue for that request and I can try to dedicate a little time to at least look into it.

**Comment by vuhe** [comment]
Thank you for your reply.

I use info.toml to place metadata information in order to unify the writing of all ComicInfo.xml in the folder, because for webtoon, all information is the same except for the number of episodes, and it is very time-consuming and difficult to manage to unify the writing.

For libraries, it is possible to import metadata information by reading series.json files created by Mylar or special metadata files for Stump, which is useful for libraries that do not want to create a fixed file structure.

The existing metadata specification file cannot automatically set the vertical continuity of webtoon, so creating a Stump metadata file specification may be useful, or some fields can be extended on the basis of series.json.

**Comment by keramblock** [comment]
I think it is important to have an ability to modify epub metadata for such type of a library. 

P.S. Right now one of things that stops me from complete migration to stump is that sometimes my db just fails(have no idea why and when and that itself is not an issue for me). And in a such event I rely purely on books metadata to recreate it from scratch. So I am forced to use calibre to redact metadata first and then upload it to stump lib.

**Comment by aaronleopold** [comment]
> I think it is important to have an ability to modify epub metadata for such type of a library

Support for basic metadata editing was added to `unstable` already. A future enhancement is planned for "writing back" the edits to the underlying file. This is largely separate from what this feature request encapsulates, though.

I'm not sure what you mean about the database failing, but if you feel it is a specific problem with Stump please open a separate issue with lots of debugging information.

**Comment by keramblock** [comment]
> I'm not sure what you mean about the database failing, but if you feel it is a specific problem with Stump please open a separate issue with lots of debugging information.

tbh most probably that is on my side and I don't want to bother you with that until I will be sure.

**Comment by ajselzilic** [comment]
I've been reading through #254, #697, and discussion #1276, and had an idea for how metadata-based libraries could work.

This could be a third library pattern alongside the existing two:

- `SERIES_BASED`
- `COLLECTION_BASED`
- `METADATA_BASED`

For a metadata-based library, Stump would scan for books recursively without treating directories as series. After reading a book's metadata:

- A non-empty `series` field would determine its Stump series.
- `number` or `series_index` would determine its position in that series.
- A book without series metadata would remain standalone instead of being placed in a series named after its directory or the library.
- Changing the stored series metadata, either through a rescan or metadata editing, would update the book's series membership.
- Files would not be moved or rewritten.

Metadata processing would be required for this pattern. Stump already reads the relevant fields from ComicInfo.xml and OPF metadata through work such as #607 and #735.

The other two library patterns would continue working as they do now. Like the existing patterns described in #43, this choice could be made when creating a library. Changing an existing library's pattern would require a separate migration process because it could affect reading progress and other existing relationships.

The main complication is how books currently relate to libraries. A book currently belongs to its library through its series:

`media.series_id -> series.library_id`

That does not work for a standalone book because it has no real series. One possible approach would be to add `library_id` directly to `media`, so every book belongs to a library while `series_id` becomes genuinely optional. When a series is assigned, Stump would also verify that it belongs to the same library.

Metadata-created series also do not have one meaningful filesystem path because their books may be spread across several directories. Their path may therefore need to be optional, with an origin field distinguishing filesystem and metadata-created series. Path-based features such as the Files tab and series-level filesystem scans would not apply to metadata-created series.

Series could be matched within a library using the trimmed metadata name while retaining their normal generated database IDs. Nested series, multiple collection memberships, story arcs, and file organization are separate concepts covered by their existing discussions. In particular, the structured collection work in #1323 can remain separate from the one canonical Stump series assigned here.

**Comment by aaronleopold** [comment]
Hey! Yeah that's similar to some of the notes I've been dumping over the last few years, and differences mostly in small details.

I probably could have shared them more often but sometimes feel like until I've fully sorted it out people probably don't care lol That said, in case you are coming from a contributor's perspective (since you made a PR earlier today): 

<details>

<summary> Here are a few things I think are worthy of sharing from my personal notes</summary>

- I originally thought to have `organization_mode` on library configs (enum{metadata, filesystem})
- I considered a non-unit enum for organization mode to merge it with `library_pattern` but historically gql does not like that so instead figured a deprecation and refactor would be safer
- A `series` would have a `kind` field (enum{filesystem, virtual})
- The scanner would need major reworking, at least in the sense that walking a tree is very different than flattening it and manually constructing from metadata (and it would need to do both)
  - Similar to why I added `tag_cache.rs`, performance would also need to be considered to reduce db round trips
- The `path` of a series is a currently very fixed concept right now that needs handling, since a virtual one would not have a path. Ideally it would be safely nullable

</details>

To respond to a few of your points though:

> The main complication is how books currently relate to libraries

Yep, there should even be a few scattered TODOs in the code to add that relation and skip the hop. This is not the only area where it has complicated things.

> Their [metadata-created series] path may therefore need to be optional

Yeah agreed

> Metadata processing would be required for this pattern

Yeah, and as you pointed out it already exists to some extent. It just isn't fully fleshed out, and in particular (at least personally, for my own usage) I would eventually want to transition towards persisting some fields in their own tables instead of the currently monolithic `media_metadata`. At least for things like authors, so we can track e.g. `series_authors`/`media_authors` and properly model things like shared works etc. I guess this is largely unrelated to the organizational bits, but sharing for posterity.

***

If this were an area of interest as a contributor, let me know. There are a couple of larger refactors I was wanting to get done first (after working on oneshots and seeing some areas I wanted to clean up)

**Comment by ajselzilic** [comment]
Yeah, I’d love to work on this and help with the larger refactors. I’m also in the Discord now as @_ajso.

**Comment by aaronleopold** [comment]
Once oneshots are merged in (ideally by EOD) I'll be focused on that clean up (in [this branch](https://github.com/stumpapp/stump/tree/reorganize-core), if you're curious) and then things should be in a better state for contributions in this area. I'll post here afterwards. I just ask that contributions are properly sliced up, probably in a dedicated feature branch I can create like I've done for some of the [permissioning reworking](https://github.com/stumpapp/stump/tree/improve-permissioning), seeing as this feature for example would likely be way too much change at once for me to properly audit and review

**Comment by aaronleopold** [comment]
@ajselzilic The clean-up work I mentioned was merged, so things are generally in a better spot if you still wanted to take a look at working on this. No worries if not, let me know and I can setup a branch off of `nightly` which slices of work can go into

**Comment by ajselzilic** [comment]
@aaronleopold Sounds good, I can take a look and work on it.

**Comment by aaronleopold** [comment]
I have created https://github.com/stumpapp/stump/tree/metadata-based-libraries, will try to regularly keep it in sync with nightly
